client/internal/httpsender.go

```diff
@@ -98,6 +98,19 @@ func (h *HTTPSender) Run(
 	h.callbacks = callbacks
 	h.receiveProcessor = newReceivedProcessor(h.logger, callbacks, h, clientSyncedState, packagesStateProvider, capabilities, packageSyncMutex)
 
+	if callbacks != nil {
+		cs, ok := callbacks.(types.CallbacksStruct)
```

Comment:
The `CallbacksStruct` is not supposed to be the only way to implement the callbacks. What is the caller supplies a different implementation?

I think we need to find a way to make the default behavior work without relying on the fact that CallbacksStruct is a struct and may have a nil CheckRedirectFunc field.

Reply:
What can be done alternatively is to remove the conditional check here and unconditionally call the CheckRedirect callback. The `CallbacksStruct` can do its own nil check instead, and use the default `net/http` redirect policy. However, this has a side effect of placing the burden for setting redirect policy on any implementer of the type, which is going to break any existing implementations in a way that is probably worse than just requiring another method to return nil on.

I think this is kind of an inherent problem in the callbacks API. It's not possible to know if a callback has been applied, so all callbacks need to be called unconditionally.

Reply:
@echlebek I agree. I created a draft PR that gets rid of Callbacks interface in favour of the struct and allows default behaviors to be defined per callback func. Please take a look: https://github.com/open-telemetry/opamp-go/pull/324